lib/r10k/puppetfile.rb

```diff
@@ -67,142 +70,118 @@ def initialize(basedir, options_or_moduledir = nil, deprecated_path_arg = nil, d
     @moduledir       = deprecated_moduledir_arg || options.delete(:moduledir)       || File.join(basedir, 'modules')
     @puppetfile_name = deprecated_name_arg      || options.delete(:puppetfile_name) || 'Puppetfile'
     @puppetfile_path = deprecated_path_arg      || options.delete(:puppetfile_path) || File.join(basedir, @puppetfile_name)
+    @environment     = options.delete(:environment)
 
     @overrides       = options.delete(:overrides) || {}
+    @default_branch_override = @overrides.dig(:environments, :default_branch_override)
 
     logger.info _("Using Puppetfile '%{puppetfile}'") % {puppetfile: @puppetfile_path}
 
-    @modules = []
-    @managed_content = {}
     @forge   = 'forgeapi.puppetlabs.com'
 
+    @loader = ::R10K::ModuleLoader::Puppetfile.new(
+      basedir: @basedir,
+      moduledir: @moduledir,
+      puppetfile: @puppetfile_path,
+      forge: @forge,
+      overrides: @overrides,
+      environment: @environment
+    )
+
+    @loaded_content = {
+      modules: [],
+      managed_directories: [],
+      desired_contents: [],
+      purge_exclusions: []
+    }
+
     @loaded = false
   end
 
+  # @param [String] default_branch_override The default branch to use
+  #   instead of one specified in the module declaration, if applicable.
+  #   Deprecated, use R10K::ModuleLoader::Puppetfile directly and pass
+  #   the default_branch_override as an option on initialization.
   def load(default_branch_override = nil)
-    return true if self.loaded?
-    if File.readable? @puppetfile_path
-      self.load!(default_branch_override)
+    if self.loaded?
+      return @loaded_content
     else
-      logger.debug _("Puppetfile %{path} missing or unreadable") % {path: @puppetfile_path.inspect}
+      self.load!(default_branch_override)
     end
   end
 
+  # @param [String] default_branch_override The default branch to use
+  #   instead of one specified in the module declaration, if applicable.
+  #   Deprecated, use R10K::ModuleLoader::Puppetfile directly and pass
+  #   the default_branch_override as an option on initialization.
   def load!(default_branch_override = nil)
-    @default_branch_override = default_branch_override
+    if default_branch_override && (default_branch_override != @default_branch_override)
+      logger.warn("Mismatch between passed and initialized default branch overrides, preferring passed value.")
+      @loader.default_branch_override = default_branch_override
```

Comment:
No, that makes sense. Really my reply was "I don't care if it's not in the best spot because I want to deprecate and remove it when I can anyways". I updated the code earlier to try to be more clear in my intentions but I'll update the location in the meantime.